CONTRIBUTING.md

```diff
@@ -576,6 +575,60 @@ that are part of this project but not production code. The canonical example
 of this is `junit`.</dd>
 </dl>
 
+
+Reviewing and accepting your contribution
+-----------------------------------------
+
+We review every contribution carefully to ensure that the change is of high
+quality and fits well with the rest of the Elasticsearch codebase. If accepted,
+we will merge your change and usually take care of backporting it to
+appropriate branches ourselves.
+
+We really appreciate everyone who is interested in contributing to
+Elasticsearch and regret that we sometimes have to reject contributions even
+when they might appear to make genuine improvements to the system. Reviewing
+contributions can be a very time-consuming task, yet the review team is small
+and our time is very limited. In some cases the time we would need to spend on
+reviews would outweigh the benefits of a change by a significant margin.
+
+We strongly recommend you discuss your change in a Github issue before spending
+much time on its implementation. We may have to reject your contribution if it
+takes the wrong approach to solving a problem or if it solves a problem that
+does not need solving. We do not enjoy seeing time wasted like that.
+
+We often reject contributions due to the low quality of the submission since
+low-quality submissions tend to take unreasonable effort to review properly.
+Quality is rather subjective so it is hard to describe exactly how to avoid
+this, but there are some basic steps you can take to reduce the chances of
+rejection. Follow the guidelines listed above when preparing your changes. Your
+PR must at least pass the test suite and you must add tests that correspond
+with your changes too. The code must be formatted neatly and must not include
+unnecessary extra changes since these can consume a lot of extra effort during
+reviews.
+
+We sometimes reject contributions if we find ourselves performing a lot of
+review iterations without making significant progress. This usually indicates
```

Comment:
Something like "so long as you always work with us in good faith we'll work with you" would go a long way to making me feel better about saying this. There aren't really rules here and so long as you are really are trying we're not going to walk away from you.

Reply:
I'm trying to say the opposite here: to protect our own time we might indeed decide to walk away, even if the contributor is trying their hardest, but we're simply not making progress equivalent to the time we're investing.

I think this paragraph is the hardest part of these docs: it's a really tricky balance between inclusiveness and reserving the right to walk away sometimes.